File: crates/md-tmpl-typescript/src/tests/conformance.test.ts
/**
 * Cross-language conformance harness (TypeScript side).
 *
 * Replays the shared JSON corpus in `<repo>/conformance` through the
 * TypeScript `md-tmpl` engine and asserts that every case matches the recorded
 * expectation. The exact same corpus is replayed by the Rust harness
 * (`crates/md-tmpl-core/tests/conformance.rs`); if both pass, the two backends
 * are behaviourally identical on the covered surface.
 *
 * The corpus expectations were originally derived by executing this reference
 * implementation, so the TypeScript side is expected to be an exact match on
 * every case (including phase for error cases).
 */

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

import { Template } from "../index.js";

// Compiled location is dist/tests/, so the repo root is four levels up.
const HERE = dirname(fileURLToPath(import.meta.url));
const CORPUS_DIR = resolve(HERE, "../../../../conformance");

// Every corpus file holds a flat list of cases whose `expect.kind` selects how
// the case is checked.
const CORPUS_FILES = [
  "render.json",
  "interpolation.json",
  "frontmatter.json",
  "errors.json",
  "escapes.json",
  "comments.json",
] as const;

interface Expect {
  kind: "render" | "default" | "error";
  output?: string;
  defaults?: Record<string, unknown>;
  phase?: "compile" | "render" | "any";
  error_contains?: string;
}

interface Case {
  name: string;
  source: string;
  params?: Record<string, unknown>;
  env?: Record<string, unknown>;
  expect: Expect;
}

function loadCases(file: string): Case[] {
  const text = readFileSync(resolve(CORPUS_DIR, file), "utf8");
  return JSON.parse(text) as Case[];
}

function compile(c: Case): Template {
  return c.env !== undefined
    ? Template.fromSourceWithEnv(c.source, c.env)
    : Template.fromSource(c.source);
}

function messageOf(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

// Compile a case, capturing any compile-time error as a string instead of
// throwing, so error-phase logic can inspect both phases uniformly.
function tryCompile(c: Case): { tmpl: Template | null; err: string | null } {
  try {
    return { tmpl: compile(c), err: null };
  } catch (e) {
    return { tmpl: null, err: messageOf(e) };
  }
}

function checkRender(c: Case): void {
  assert.ok(c.expect.output !== undefined, "render case needs expect.output");
  const out = compile(c).render(c.params ?? {});
  assert.strictEqual(out, c.expect.output);
}

function checkDefault(c: Case): void {
  assert.ok(
    c.expect.defaults !== undefined,
    "default case needs expect.defaults",
  );
  const defs = compile(c).defaults();
  assert.deepStrictEqual(defs, c.expect.defaults);
}

// Assert a rendered error message contains the recorded needle, when one is
// recorded. A missing needle means "any error is acceptable".
function assertNeedle(needle: string | undefined, haystack: string): void {
  if (needle !== undefined) {
    assert.ok(
      haystack.includes(needle),
      `error ${JSON.stringify(haystack)} lacks substring ${JSON.stringify(needle)}`,
    );
  }
}

function checkError(c: Case): void {
  const phase = c.expect.phase;
  assert.ok(phase !== undefined, "error case needs expect.phase");
  const needle = c.expect.error_contains;
  const { tmpl, err } = tryCompile(c);

  if (phase === "compile") {
    assert.ok(err !== null, "expected a COMPILE error but compile succeeded");
    assertNeedle(needle, err);
    return;
  }

  // Both "render" and "any" require a successful compile before rendering,
  // except "any" also accepts a compile-time failure (leak-safety may trip at
  // either phase; the phase is allowed to differ between backends).
  if (err !== null) {
    assert.strictEqual(
      phase,
      "any",
      `expected a RENDER error but failed at COMPILE: ${err}`,
    );
    assertNeedle(needle, err);
    return;
  }

  assert.ok(tmpl !== null, "compile reported success but produced no template");
  let renderErr: string | null = null;
  try {
    tmpl.render(c.params ?? {});
  } catch (e) {
    renderErr = messageOf(e);
  }
  assert.ok(renderErr !== null, "expected a RENDER error but render succeeded");
  assertNeedle(needle, renderErr);
}

function runCase(c: Case): void {
  switch (c.expect.kind) {
    case "render":
      checkRender(c);
      break;
    case "default":
      checkDefault(c);
      break;
    case "error":
      checkError(c);
      break;
    default:
      throw new Error(`unknown expect.kind for case ${c.name}`);
  }
}

for (const file of CORPUS_FILES) {
  describe(`conformance: ${file}`, () => {
    const cases = loadCases(file);
    assert.ok(cases.length > 0, `corpus file ${file} is empty`);
    for (const c of cases) {
      it(c.name, () => {
        runCase(c);
      });
    }
  });
}

File: go/md_tmpl/conformance_test.go
package md_tmpl

// Cross-language conformance harness (Go side).
//
// Replays the shared TOML corpus in <repo>/tests/conformance through the Go
// md_tmpl engine and asserts that every case matches the recorded expectation.
// The exact same corpus is replayed by the Rust, TypeScript, and Python
// harnesses; if all pass, the four backends are behaviourally identical on the
// covered surface.
//
// TOML has no null, so option-None is encoded in the corpus as the sentinel
// inline table `{ __none__ = true }` and decoded back to Go nil on load.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	"github.com/BurntSushi/toml"
)

func corpusDir() string {
	// go/md_tmpl -> go -> <repo root>.
	return filepath.Join("..", "..", "tests", "conformance")
}

var corpusFiles = []string{
	"render.toml",
	"interpolation.toml",
	"frontmatter.toml",
	"errors.toml",
	"escapes.toml",
	"comments.toml",
	"literals.toml",
}

// denull decodes the corpus's `{ __none__ = true }` option-None sentinel back
// into a Go nil (TOML has no null of its own).
func denull(v any) any {
	switch x := v.(type) {
	case []map[string]any:
		for i := range x {
			if m, ok := denull(x[i]).(map[string]any); ok {
				x[i] = m
			}
		}
		return x
	case []any:
		for i := range x {
			x[i] = denull(x[i])
		}
		return x
	case map[string]any:
		if len(x) == 1 {
			if b, ok := x["__none__"].(bool); ok && b {
				return nil
			}
		}
		for k := range x {
			x[k] = denull(x[k])
		}
		return x
	default:
		return v
	}
}

func loadCases(t *testing.T, file string) []map[string]any {
	t.Helper()
	path := filepath.Join(corpusDir(), file)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read corpus file %s: %v", path, err)
	}
	var root map[string]any
	if err := toml.Unmarshal(data, &root); err != nil {
		t.Fatalf("parse corpus file %s: %v", path, err)
	}

	var cases []map[string]any
	switch arr := root["cases"].(type) {
	case []map[string]any:
		cases = arr
	case []any:
		for _, item := range arr {
			m, ok := item.(map[string]any)
			if !ok {
				t.Fatalf("%s: a case is not a table (%T)", file, item)
			}
			cases = append(cases, m)
		}
	default:
		t.Fatalf("%s: [[cases]] has unexpected type %T", file, root["cases"])
	}

	for i := range cases {
		if m, ok := denull(cases[i]).(map[string]any); ok {
			cases[i] = m
		}
	}
	return cases
}

// jsonNorm projects a value through JSON so the Go engine's numeric types
// (int64/int/float64) and the TOML-decoded types compare structurally.
func jsonNorm(t *testing.T, v any) any {
	t.Helper()
	b, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("normalize marshal: %v", err)
	}
	var out any
	if err := json.Unmarshal(b, &out); err != nil {
		t.Fatalf("normalize unmarshal: %v", err)
	}
	return out
}

func compileCase(c map[string]any) (*Template, error) {
	source, _ := c["source"].(string)
	if env, ok := c["env"].(map[string]any); ok {
		return FromSourceWithEnv(source, env)
	}
	return FromSource(source)
}

func caseParams(c map[string]any) map[string]any {
	if params, ok := c["params"].(map[string]any); ok {
		return params
	}
	return map[string]any{}
}

func assertNeedle(t *testing.T, needle, haystack string) {
	t.Helper()
	if needle != "" && !strings.Contains(haystack, needle) {
		t.Fatalf("error %q lacks substring %q", haystack, needle)
	}
}

func checkRender(t *testing.T, c map[string]any, expect map[string]any) {
	t.Helper()
	tmpl, err := compileCase(c)
	if err != nil {
		t.Fatalf("compile error: %v", err)
	}
	got, err := tmpl.RenderMap(caseParams(c))
	if err != nil {
		t.Fatalf("render error: %v", err)
	}
	want, _ := expect["output"].(string)
	if got != want {
		t.Fatalf("render mismatch\n  want: %q\n  got:  %q", want, got)
	}
}

func checkDefault(t *testing.T, c map[string]any, expect map[string]any) {
	t.Helper()
	tmpl, err := compileCase(c)
	if err != nil {
		t.Fatalf("compile error: %v", err)
	}
	want, _ := expect["defaults"].(map[string]any)
	if want == nil {
		want = map[string]any{}
	}
	got := tmpl.Defaults()
	if !reflect.DeepEqual(jsonNorm(t, got), jsonNorm(t, want)) {
		t.Fatalf("defaults mismatch\n  want: %v\n  got:  %v", want, got)
	}
}

func checkError(t *testing.T, c map[string]any, expect map[string]any) {
	t.Helper()
	phase, _ := expect["phase"].(string)
	needle, _ := expect["error_contains"].(string)
	tmpl, compileErr := compileCase(c)

	switch phase {
	case "compile":
		if compileErr == nil {
			t.Fatalf("expected a COMPILE error but compile succeeded")
		}
		assertNeedle(t, needle, compileErr.Error())
	case "render", "any":
		// Both require a successful compile before rendering, except "any" also
		// accepts a compile-time failure (leak-safety may trip at either phase;
		// the phase is allowed to differ between backends).
		if compileErr != nil {
			if phase != "any" {
				t.Fatalf("expected a RENDER error but failed at COMPILE: %v", compileErr)
			}
			assertNeedle(t, needle, compileErr.Error())
			return
		}
		if _, err := tmpl.RenderMap(caseParams(c)); err != nil {
			assertNeedle(t, needle, err.Error())
		} else {
			t.Fatalf("expected a RENDER error but render succeeded")
		}
	default:
		t.Fatalf("bad phase %q", phase)
	}
}

func TestConformance(t *testing.T) {
	for _, file := range corpusFiles {
		cases := loadCases(t, file)
		if len(cases) == 0 {
			t.Fatalf("corpus file %s is empty", file)
		}
		for _, c := range cases {
			name, _ := c["name"].(string)
			expect, ok := c["expect"].(map[string]any)
			if !ok {
				t.Fatalf("%s/%s: missing expect table", file, name)
			}
			t.Run(file+"/"+name, func(t *testing.T) {
				switch expect["kind"] {
				case "render":
					checkRender(t, c, expect)
				case "default":
					checkDefault(t, c, expect)
				case "error":
					checkError(t, c, expect)
				default:
					t.Fatalf("unknown expect.kind %v", expect["kind"])
				}
			})
		}
	}
}
